src/coreclr/jit/codegenxarch.cpp

```diff
@@ -7331,6 +7331,18 @@ void CodeGen::genIntToFloatCast(GenTree* treeNode)
     noway_assert(srcType != TYP_UINT);
     noway_assert((srcType != TYP_ULONG) || (dstType != TYP_FLOAT));
 
+    if (compiler->compOpportunisticallyDependsOn(InstructionSet_AVX512F))
+    {
+        if (srcType == TYP_ULONG && (dstType == TYP_DOUBLE || dstType == TYP_FLOAT))
```

Comment:
We still have a comment and noway assert above that is blocking `srcType == TYP_ULONG && dstType == TYP_FLOAT`, so we're either going to trigger an assert or not hitting this path as expected for that scenario

Reply:
Yes. We need that assert for now because we have not optimized the case of ulong->float. This is because of the following scenario:
```
 * GT_CAST(uint64, float)          =   GT_CAST(GT_CAST(uint64, double), float)
```
Hence the above assert where we are casting from ulong to float will not be hit since it is broken down into 2 casts on import.

Reply:
We should likely drop the unreachable part here then, and only introduce it when the support is added. That is
```suggestion
        if ((srcType == TYP_ULONG) && (dstType == TYP_DOUBLE))
```

Reply:
Tweaking import to generate `GT_CAST(uint64, float)` directly is also desirable, but we can do that in a follow up PR.

Reply:
On second thought, given all the changes we have around `vcvtss2usi` in this PR, I'd expect us to be handling this as well.

We don't want to be adding code that is untested or actively being unused if we can help it.

The PR, as it is, has handling for emitting `vcvttusi2sd`. It also has changes around but does not actually emit `vcvttusi2ss`

Reply:
The minor fixes to `vcvttss2usi` so that its split as `vcvttss2usi32/64` seem fine provided they are fixing an issue impacting the PR itself. Otherwise, it would be more desirable to have them in their own PR (ideally alongside the support to emit them as part of general scalar casts).

Reply:
@tannergooding are you saying we tweak the importer to import as `GT_CAST(uint64, float)` and then optimize that case?

As opposed to crunching `GT_CAST(GT_CAST(uint64, double), float)` into `GT_CAST(uint64, double)` which is a little challenging (though not undoable) I think due to the way morph tries to expand casts.

Reply:
I agree. We will rework the PR to handle the additional cases.

Reply:
I am away for 2 weeks and will be back on 06/12. Will work on this once I am back.

Reply:
@tannergooding I think the case that is a little bit trickier is this...

```C#
public static float M2(ulong v) {
  return (float)v;
}
```

becomes this in the IL...
'

```
IL_0000: ldarg.0
IL_0001: conv.r.un
IL_0002: conv.r4
IL_0003: ret
```

which means the importer isn't expanding a cast here, it is already split from its input. I think this would require I little bit of extra handling in the importer itself?

Reply:
> 

 I have added a handling for the following with the latest changes

1. Merge GT_CAST(GT_CAST(TYP_ULONG, TYP_DOUBLE), TYP_FLOAT) into GT_CAST(TYP_ULONG, TYP_FLOAT)
2. Now no generating a large node for 
3. Changes already present to not use helper function in case of AVX512 being present